### PR TITLE
Fix support for the bfd library

### PR DESCRIPTION
**Pull request by shindere**
This PR is a tentative fix for issue #9297.

There was a regression: -ldl was not tried to link with the bfd library.

This PR fixes this regression by re-introducing "-ldl".

@Octachron: if this works, it would be good to include it in 4.10, I think.

### CODE REVIEW

**Comment by olafhering** [comment]
It is actually `-ldL`

**Comment by shindere** [comment]
Olaf Hering (2020/02/12 06:49 -0800):
> It is actually `-ldL`

Sorry? You really mean the uppercase L?

I checked and I used '-ldl' (lower-case) and it was that option which
was used before autoconf was introduced.

**Comment by olafhering** [comment]
I was unclear, sorry. Just the description is incorrect, 

The patch works, it succeeds with the `-lbfd -ldl -liberty -lz -lm` case.

**Comment by shindere** [comment]
Olaf Hering (2020/02/12 15:08 +0000):
> I was unclear, sorry. Just the description is incorrect,

Ah, sorry. I fixed the commit log but not yet the one in the PR itself.
Would you mind doing it, @gasche?

> The patch works, it succeeds with the `-lbfd -ldl -liberty -lz -lm`
> case.

That's good to know! Thanks!

And thanks, too, for your reactivity on this!

**Review by dra27 — APPROVED** [review]
Just needs a `Changes` entry.

**Comment by shindere** [comment]
David Allsopp (2020/02/13 08:08 -0800):
> dra27 approved this pull request.

Thank you!

> Just needs a `Changes` entry.

Done, added this PR to the existing one and you as a reviewer, @dra27.
Also rebased.

**Comment by Octachron** [comment]
Concerning the inclusion in 4.10, since this is a regression fix on a configure option, I am of for inclusion in 4.10 . I will merge and cherry-pick once the Ci is green.

**Comment by olafhering** [comment]
This regression exists also in 4.08 and 4.09. Is this a backport candidate?

**Comment by shindere** [comment]
Olaf Hering (2020/02/13 11:03 -0800):
> This regression exists also in 4.08 and 4.09.

Yes, we are aware and I'm really sorry about that.

> Is this a backport candidate?

I don't think there will be any bugfix release upstream but then of
course package maintainers can provide whatever they feel is useful to
their end-users.

**Comment by Octachron** [comment]
There is a 4.09.1 maintenance release planned once I am done with 4.10 . If there is some indication that a 4.08.2 release would help downstream package manager, I am willing to consider it.

**Comment by shindere** [comment]
Plenty of thanks, @Octachron, you are doing a wonderful job!

**Comment by shindere** [comment]
Florian Angeletti (2020/02/14 01:15 -0800):
> Merged #9302 into trunk.

Thanks!

**Comment by shindere** [comment]
Okay, we now have again a problem on OpenBSD where `-ldl` is not
recognized.  Concretely, it means that configure won't be able to
detect BFD support properly on OpenBSD, which will cause an error in
presence of the `--with-bfd` configure option.

I am not sure this is a real regression as compared to the pre-autoconf
situation, though, because, as far as I know, nothing confirms that this
actually worked before.

I don't know yet how this should best be fixed.

Will think about it, but any suggestion welcome.

One possibility would be to try `-ldl` only if the OS is different from
OpenBSD. It wouldn't be too difficult to implement, but not sure that's the
right way to go?

**Comment by gasche** [comment]
As someone with no clue about how autoconf works, the intuitive idea would be to first try without `-ldl` and then try with it, and record in a variable somewhere the flags that actually worked (if any), for use when compiling the actual code that needs the bfd library.

**Comment by olafhering** [comment]
Which `AC_CHECK_LIB` did succeed before on openBSD? I think just adding the final newly added `AC_CHECK_LIB` call to cover binutils with `--enable-plugns` should be enough to avoid regressions. Then the added `-ldl` can be removed again from the existing `AC_CHECK_LIB` calls.

**Comment by shindere** [comment]
@olafhering: I just opened #9311. Could you please check that it still
works for you?

**Pull request merged**
